File: wakakusa/demo/canvas.rb
# A canvas: a grid of virtual pixels, painted command by command.
#
# `canvas(width, height, scale:, background:, palette:)` opens the grid
# and its block paints — pixel, line, rect, rect_outline, circle,
# circle_outline, triangle, triangle_outline, sprite and pixel_text.
# `scale` says how many logical pixels one virtual pixel takes, so a
# 64x40 canvas at six is 384x240 on screen.
#
# Every color is a NUMBER: the index of a color in `palette`. That is
# how tools for pixel art work, so drawing code written for one moves
# here with its numbers unchanged.
#
# The commands are not elements. Nothing here can be clicked, themed,
# sized or animated, and a loop inside the canvas is the ordinary loop:
# what its body paints joins the frame where it stands.
require "wakakusa"

HEADING = { size: 18.0, color: "accent" }.freeze
FAINT = { size: 12.0, color: "#8a8f98" }.freeze

# Five colors are enough to show that the index IS the color.
PALETTE = ["#11111b", "#89b4fa", "#f38ba8", "#eeeeee", "#a6e3a1"].freeze

class Blip
  attr_reader :x, :y, :c

  def initialize(x, y, c)
    @x = x
    @y = y
    @c = c
  end
end

class Sky
  def initialize
    @frame = 0
    @ball_x = 30
    @ball_y = 18
    @dx = 1
    @dy = 1
    @blips = []
    seed
  end

  def seed
    @blips = [Blip.new(6, 4, 1), Blip.new(20, 9, 2), Blip.new(50, 6, 3), Blip.new(58, 30, 4)]
  end

  def tick
    @frame += 1
    # The keyboard is read here, in the tick, never in a view.
    # `key_down` is "held right now", so holding an arrow steers.
    @dx = -1 if key_down("left")
    @dx = 1 if key_down("right")
    @dy = -@dy if key_pressed("space")
    x = @ball_x + @dx
    y = @ball_y + @dy
    if x < 4
      x = 4
      @dx = 1
    end
    if x > 59
      x = 59
      @dx = -1
    end
    if y < 4
      y = 4
      @dy = 1
    end
    if y > 35
      y = 35
      @dy = -1
    end
    @ball_x = x
    @ball_y = y
  end

  def blip(b)
    pixel(b.x, b.y, b.c)
  end

  def view
    column(spacing: 12.0, padding: 16.0) {
      text "Canvas", **HEADING
      text "a grid of virtual pixels; every color is an index", **FAINT
      canvas(64, 40, scale: 6, background: 0, palette: PALETTE) {
        rect(2, 2, 12, 6, 1)
        rect_outline(16, 2, 12, 6, 2)
        circle_outline(34, 5, 4, 3)
        line(2, 11, 61, 11, 2)
        triangle(3, 37, 8, 28, 13, 37, 4)
        @blips.each { |b| blip(b) }
        circle(@ball_x, @ball_y, 3, 3)
        pixel_text(2, 14, "FRAME #{@frame}", 3)
      }
      row(spacing: 8.0) {
        button("seed") { seed }
      }
    }
  end
end

app = Sky.new
every(0.05) { app.tick }
run(app, title: "canvas")

File: gomamochi/demo/canvas.go
// A canvas: a grid of virtual pixels, painted command by command.
//
// `Canvas(width, height)` opens the grid; `Scale` says how many logical
// pixels one virtual pixel takes, so a 64x40 canvas at six is 384x240
// on screen; and `Paint` is handed a painter with the commands — Pixel,
// Line, Rect, RectOutline, Circle, CircleOutline, Triangle,
// TriangleOutline, Sprite and PixelText.
//
// Every color is a NUMBER: the index of a color in `Palette`. That is
// how tools for pixel art work, so drawing code written for one moves
// here with its numbers unchanged.
//
// The commands are not elements. Nothing here can be clicked, themed,
// sized or animated, and a loop inside the paint closure is the
// ordinary loop: what its body paints joins the frame where it stands.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

// Five colors are enough to show that the index IS the color.
var palette = []string{"#11111b", "#89b4fa", "#f38ba8", "#eeeeee", "#a6e3a1"}

type Blip struct {
	x, y, c int
}

type Sky struct {
	frame  int
	ballX  int
	ballY  int
	dx, dy int
	blips  []Blip
}

func (s *Sky) seed() {
	s.blips = []Blip{{6, 4, 1}, {20, 9, 2}, {50, 6, 3}, {58, 30, 4}}
}

func (s *Sky) tick() {
	s.frame += 1
	// The keyboard is read here, in the tick, never in a view.
	// KeyDown is "held right now", so holding an arrow steers.
	if KeyDown("left") {
		s.dx = -1
	}
	if KeyDown("right") {
		s.dx = 1
	}
	if KeyPressed("space") {
		s.dy = -s.dy
	}
	x := s.ballX + s.dx
	y := s.ballY + s.dy
	if x < 4 {
		x = 4
		s.dx = 1
	}
	if x > 59 {
		x = 59
		s.dx = -1
	}
	if y < 4 {
		y = 4
		s.dy = 1
	}
	if y > 35 {
		y = 35
		s.dy = -1
	}
	s.ballX = x
	s.ballY = y
}

func (s *Sky) View() Element {
	return Column(
		Text("Canvas").Size(18).Color("accent"),
		Text("a grid of virtual pixels; every color is an index").Size(12).Color("#8a8f98"),
		Canvas(64, 40).Scale(6).Background(0).Palette(palette...).Paint(func(p *Painter) {
			p.Rect(2, 2, 12, 6, 1)
			p.RectOutline(16, 2, 12, 6, 2)
			p.CircleOutline(34, 5, 4, 3)
			p.Line(2, 11, 61, 11, 2)
			p.Triangle(3, 37, 8, 28, 13, 37, 4)
			for _, b := range s.blips {
				p.Pixel(b.x, b.y, b.c)
			}
			p.Circle(s.ballX, s.ballY, 3, 3)
			p.PixelText(2, 14, fmt.Sprintf("FRAME %d", s.frame), 3)
		}),
		Row(
			Button("seed").OnClick(func() { s.seed() }),
		).Spacing(8),
	).Spacing(12).Padding(16)
}

func main() {
	app := &Sky{ballX: 30, ballY: 18, dx: 1, dy: 1}
	app.seed()
	Every(0.05, func() { app.tick() })
	Run(app, Title("canvas"))
}
